app/src/main/cpp/skyline/kernel/ipc.h

```diff
@@ -1,328 +1,337 @@
+            u8 inputCount;
+            u16 payloadSz;
+            u32 objectId;
+            u32 : 32;
+            u32 token;
+        };
+        static_assert(sizeof(DomainHeaderRequest) == 16);
 
         /**
-         * @param address The address of the buffer
-         * @param size The size of the buffer
-         * @param type The type of the buffer
+         * @brief This reflects the value of DomainHeaderRequest::command
          */
-        IpcBuffer(u64 address, size_t size, IpcBufferType type);
-    };
+        enum class DomainCommand : u8 {
+            SendMessage = 1, CloseVHandle = 2
+        };
 
-    /**
-     * @brief This holds an input IPC buffer
-     */
-    struct InputBuffer : public IpcBuffer {
         /**
-         * @param aBuf The X Buffer Descriptor that has contains the input data
+         * @brief This bit-field structure holds the domain's header of an IPC response command. (https://switchbrew.org/wiki/IPC_Marshalling#Domains)
          */
-        InputBuffer(kernel::ipc::BufferDescriptorX *xBuf);
+        struct DomainHeaderResponse {
+            u32 outputCount;
+            u32 : 32;
+            u64 : 64;
+        };
+        static_assert(sizeof(DomainHeaderResponse) == 16);
 
         /**
-         * @param aBuf The A or W Buffer Descriptor that has contains the input data
+         * @brief This bit-field structure holds the data payload of an IPC command. (https://switchbrew.org/wiki/IPC_Marshalling#Data_payload)
          */
-        InputBuffer(kernel::ipc::BufferDescriptorABW *aBuf, IpcBufferType type = IpcBufferType::A);
-    };
+        struct PayloadHeader {
+            u32 magic;
+            u32 version;
+            u32 value;
+            u32 token;
+        };
+        static_assert(sizeof(PayloadHeader) == 16);
+
+
+        /**
+         * @brief This reflects which function PayloadHeader::value refers to when a control request is sent (https://switchbrew.org/wiki/IPC_Marshalling#Control)
+         */
+        enum class ControlCommand : u32 {
+            ConvertCurrentObjectToDomain = 0, CopyFromCurrentDomain = 1, CloneCurrentObject = 2, QueryPointerBufferSize = 3, CloneCurrentObjectEx = 4
+        };
+
+        /**
+         * @brief This is a buffer descriptor for X buffers: https://switchbrew.org/wiki/IPC_Marshalling#Buffer_descriptor_X_.22Pointer.22
+         */
+        struct BufferDescriptorX {
+            u16 counter0_5   : 6; //!< The first 5 bits of the counter
+            u16 address36_38 : 3; //!< Bit 36-38 of the address
+            u16 counter9_11  : 3; //!< Bit 9-11 of the counter
+            u16 address32_35 : 4; //!< Bit 32-35 of the address
+            u16 size         : 16; //!< The 16 bit size of the buffer
+            u32 address0_31  : 32; //!< The first 32-bits of the address
+
+            BufferDescriptorX(u64 address, u16 counter, u16 size) : size(size) {
+                address0_31 = static_cast<u32>(address & 0x7FFFFFFF80000000);
```

Comment:
Might me an idea to make some constexpr function that generates a mask like linux's [GENMASK](https://elixir.bootlin.com/linux/v4.4/source/include/linux/bitops.h#L21) and const for params

Reply:
I don't think it'll be used all that much because we generally use bitfields for that and values aren't broken up like this a lot either way, in fact even here the constructor isn't really used. It's just present for completeness's sake.